clamd/thrmgr.c

```diff
@@ -700,10 +700,10 @@ static void *thrmgr_worker(void *arg)
             exit(-2);
         }
         if (job_data) {
-            threadpool->handler(job_data);
-        } else if (must_exit) {
-            break;
-        }
+            if (must_exit) {
+                break;
+            }                
+        } 
```

Comment:
I'm not sure I understand the issue you're trying to fix. But this doesn't look like a correct solution.  I'm pretty sure we need to call the `handler()` function.  

And the `if (job_data)` condition exists so we only call it if the `job_data` is defined. If it weren't needed, then I think you could change it to just:
```c
        if (must_exit) {
            break;
        }
```
but since I think it is needed -- maybe the change you were looking for is this?:
```c
        if (job_data) {
            threadpool->handler(job_data);
        }
        if (must_exit) {
            break;        
        }
```

or maybe this?:
```c
        if (must_exit) {
            break;        
        }
        if (job_data) {
            threadpool->handler(job_data);
        }
```

I'm not recommending these changes.  I haven't tested anything yet.  I'm still not sure what the problem is or how to test that it is fixed.

Can you provided steps for how to reproduce the problem so I can test these possible fixes?

Reply:
After submit this pull and leave system run more and check I found that it is not fix process thread when client terminate connection or timeout limit. I still can not find how to terminate or cancel scan thread when it already got ETIMEDOUT. There should have some control function to stop thread immediately but I could not found one or I am not understand it because I am not good in C language.